Wold-style-cast

**Describe the bug**
There are several old style casts caught by g++:

```
./catch_amalgamated.hpp: In function ‘Catch::Benchmark::Estimate<double> Catch::Benchmark::Detail::bootstrap(double, Iterator, Iterator, const sample&, Estimator&&)’:
./catch_amalgamated.hpp:2366:132: error: use of old-style cast to ‘double’ [-Werror=old-style-cast]
 2366 |                 double prob_n = std::count_if(resample.begin(), resample.end(), [point](double x) { return x < point; }) / (double)n;
      |                                                                                                                                    ^
./catch_amalgamated.cpp: In lambda function:
./catch_amalgamated.cpp:193:66: error: use of old-style cast to ‘int’ [-Werror=old-style-cast]
  193 |                     return (int)(-2. * k0 / (k1 + std::sqrt(det)));
      |                                                                  ^
      |                            -----
      |                            static_cast<int> (                     )
```

**Reproduction steps**
Steps to reproduce the bug:

Enable `-Wold-style-cast` in your CI.

**Catch2 version**

```
   8   │ //  Catch v3.0.0-preview.3
   9   │ //  Generated: 2020-10-08 13:59:26.309308
```

--- a/CMake/MiscFunctions.cmake
+++ b/CMake/MiscFunctions.cmake
@@ -63,6 +63,7 @@ function(add_warnings_to_targets targets)
           "-Wdeprecated-register"
           "-Wsuggest-override"
           "-Wshadow"
+          "-Wold-style-cast"
         )
         foreach(warning ${CHECKED_WARNING_FLAGS})
             add_cxx_flag_if_supported_to_targets(${warning} "${targets}")

--- a/src/catch2/benchmark/detail/catch_stats.cpp
+++ b/src/catch2/benchmark/detail/catch_stats.cpp
@@ -182,7 +182,7 @@ namespace Catch {
                     double k0 = -n * nd;
                     double k1 = sb2 - n * sg2 + nd;
                     double det = k1 * k1 - 4 * sg2 * k0;
-                    return (int)(-2. * k0 / (k1 + std::sqrt(det)));
+                    return static_cast<int>(-2. * k0 / (k1 + std::sqrt(det)));
                 };
 
                 auto var_out = [n, sb2, sg2](double c) {

--- a/src/catch2/benchmark/detail/catch_stats.hpp
+++ b/src/catch2/benchmark/detail/catch_stats.hpp
@@ -102,7 +102,7 @@ namespace Catch {
 
                 double accel = sum_cubes / (6 * std::pow(sum_squares, 1.5));
                 int n = static_cast<int>(resample.size());
-                double prob_n = std::count_if(resample.begin(), resample.end(), [point](double x) { return x < point; }) / (double)n;
+                double prob_n = std::count_if(resample.begin(), resample.end(), [point](double x) { return x < point; }) / static_cast<double>(n);
                 // degenerate case with uniform samples
                 if (prob_n == 0) return { point, point, point, confidence_level };
 

--- a/tests/SelfTest/Baselines/compact.sw.approved.txt
+++ b/tests/SelfTest/Baselines/compact.sw.approved.txt
@@ -1342,7 +1342,7 @@ String.tests.cpp:<line number>: passed: s.data() == ss.data() for: "hello world!
 String.tests.cpp:<line number>: passed: s.substr(s.size() + 1, 123).empty() for: true
 String.tests.cpp:<line number>: passed: std::strcmp(ss.c_str(), "world!") == 0 for: 0 == 0
 String.tests.cpp:<line number>: passed: s.substr(1'000'000, 1).empty() for: true
-String.tests.cpp:<line number>: passed: (char*)buffer1 != (char*)buffer2 for: "Hello" != "Hello"
+String.tests.cpp:<line number>: passed: reinterpret_cast<char*>(buffer1) != reinterpret_cast<char*>(buffer2) for: "Hello" != "Hello"
 String.tests.cpp:<line number>: passed: left == right for: Hello == Hello
 String.tests.cpp:<line number>: passed: left != left.substr(0, 3) for: Hello != Hel
 String.tests.cpp:<line number>: passed: sr == "a standard string" for: a standard string == "a standard string"

--- a/tests/SelfTest/Baselines/console.sw.approved.txt
+++ b/tests/SelfTest/Baselines/console.sw.approved.txt
@@ -9968,7 +9968,7 @@ String.tests.cpp:<line number>
 ...............................................................................
 
 String.tests.cpp:<line number>: PASSED:
-  CHECK( (char*)buffer1 != (char*)buffer2 )
+  CHECK( reinterpret_cast<char*>(buffer1) != reinterpret_cast<char*>(buffer2) )
 with expansion:
   "Hello" != "Hello"
 

--- a/tests/SelfTest/Baselines/tap.sw.approved.txt
+++ b/tests/SelfTest/Baselines/tap.sw.approved.txt
@@ -2598,7 +2598,7 @@ ok {test-number} - std::strcmp(ss.c_str(), "world!") == 0 for: 0 == 0
 # StringRef
 ok {test-number} - s.substr(1'000'000, 1).empty() for: true
 # StringRef
-ok {test-number} - (char*)buffer1 != (char*)buffer2 for: "Hello" != "Hello"
+ok {test-number} - reinterpret_cast<char*>(buffer1) != reinterpret_cast<char*>(buffer2) for: "Hello" != "Hello"
 # StringRef
 ok {test-number} - left == right for: Hello == Hello
 # StringRef

--- a/tests/SelfTest/Baselines/xml.sw.approved.txt
+++ b/tests/SelfTest/Baselines/xml.sw.approved.txt
@@ -12133,7 +12133,7 @@ Message from section two
       <Section name="Comparisons are deep" filename="tests/<exe-name>/IntrospectiveTests/String.tests.cpp" >
         <Expression success="true" type="CHECK" filename="tests/<exe-name>/IntrospectiveTests/String.tests.cpp" >
           <Original>
-            (char*)buffer1 != (char*)buffer2
+            reinterpret_cast&lt;char*>(buffer1) != reinterpret_cast&lt;char*>(buffer2)
           </Original>
           <Expanded>
             "Hello" != "Hello"

--- a/tests/SelfTest/IntrospectiveTests/String.tests.cpp
+++ b/tests/SelfTest/IntrospectiveTests/String.tests.cpp
@@ -79,7 +79,7 @@ TEST_CASE( "StringRef", "[Strings][StringRef]" ) {
     SECTION( "Comparisons are deep" ) {
         char buffer1[] = "Hello";
         char buffer2[] = "Hello";
-        CHECK((char*)buffer1 != (char*)buffer2);
+        CHECK(reinterpret_cast<char*>(buffer1) != reinterpret_cast<char*>(buffer2));
 
         StringRef left(buffer1), right(buffer2);
         REQUIRE( left == right );

--- a/tests/SelfTest/IntrospectiveTests/Xml.tests.cpp
+++ b/tests/SelfTest/IntrospectiveTests/Xml.tests.cpp
@@ -41,7 +41,7 @@ TEST_CASE( "XmlEncode", "[XML]" ) {
 
 // Thanks to Peter Bindels (dascandy) for some of the tests
 TEST_CASE("XmlEncode: UTF-8", "[XML][UTF-8][approvals]") {
-#define ESC(lit) (char*)(lit)
+#define ESC(lit) reinterpret_cast<const char*>(lit)
     SECTION("Valid utf-8 strings") {
         CHECK(encode(ESC(u8"Here be 👾")) == ESC(u8"Here be 👾"));
         CHECK(encode(ESC(u8"šš")) == ESC(u8"šš"));

